tests/examples/shims/client.lua

```diff
@@ -0,0 +1,202 @@
+local gears_obj = require("gears.object")
+
+local clients = {}
+
+--TODO merge all other implementations
+local function convert_singleton(obj)
+    obj._connect_signal = obj.connect_signal
+    function obj.connect_signal(name, func)
+        return obj._connect_signal(obj, name, func)
+    end
+    obj.weak_connect_signal = obj.connect_signal
+
+    obj._add_signal = obj.add_signal
+
+    function obj.add_signal(name)
+        return obj._add_signal(obj, name)
+    end
+
+    function obj.set_index_miss_handler(_)
+        --TODO
+    end
+
+    function obj.set_newindex_miss_handler(_)
+        --TODO
+    end
+
+    function obj.emit_signal(name,c,...)
+        local conns = obj._signals[name]
+        for func in pairs(conns.strong) do
+            func(c, ...)
+        end
+        for func in pairs(conns.weak) do
+            func(c, ...)
+        end
+    end
+
+    return obj
+end
+
+local client = convert_singleton(gears_obj())
+
+local function add_signals(c)
+    c:add_signal("property::width")
+    c:add_signal("property::height")
+    c:add_signal("property::x")
+    c:add_signal("property::y")
+    c:add_signal("property::screen")
+    c:add_signal("property::geometry")
+    c:add_signal("request::geometry")
+    c:add_signal("swapped")
+    c:add_signal("raised")
+    c:add_signal("property::_label") --Used internally
+end
+
+-- Keep an history of the geometry for validation and images
+local function push_geometry(c)
+    table.insert(c._old_geo, c:geometry())
+end
+
+-- Undo the last move, but keep it in history
+-- local function pop_geometry(c)
+--     CURRENTLY UNUSED
+-- end
+
+-- Create fake clients to move around
+function client.gen_fake(args)
+    local ret = gears_obj()
+    ret.type = "normal"
+    ret.valid = true
+    ret.size_hints = {}
+    ret.border_width = 1
+
+    -- Apply all properties
+    for k,v in pairs(args or {}) do
+        ret[k] = v
+    end
+
+    -- Tests should always set a geometry, but just in case
+    for _, v in ipairs{"x","y","width","height"} do
+        ret[v] = ret[v] or 1
+    end
+
+
+    add_signals(ret)
+
+    -- Emulate capi.client.geometry
+    function ret:geometry(new)
+        if new then
+            for k,v in pairs(new) do
+                ret:emit_signal("property::"..k, v)
+                ret[k] = v
```

Comment:
I think the order is wrong, you should first change the value and then emit the signal (and only emit if the value actually changed)